Create ADR for WASM-compiled `cardano-api` (#75)
input-output-hk/cardano-node-wiki: docs/ADR-014-Cardano-API-WASM-library-for-browser.md

Reviewer: There's no "Decision" section. What are negative consequences of implementing this ADR?

Author: Addressed here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/8691367cfdfee2b5387249cda10827d54185edbe

# ADR-014: JavaScript API for Cardano via WASM-compiled cardano-api

**Status:** Proposed

**Date:** 2025-05-22

## Context

Decentralized Applications (DApps) are primarily utilized through web browsers. And Cardano DApps in Cardano would benefit from the extensive features provided by `cardano-api`.

> There already exist several JavaScript/TypeScript libraries and APIs that support DApps (e.g., Lucid Evolution, Helios, MeshJS, cardano-serialization-lib, cardano-sdk-js, Ogmios, TyphonJS), but porting `cardano-api` can offer advantages over other approaches:

Reviewer: Could you turn those names to links?

> * **Comprehensiveness:** `cardano-api` is especially comprehensive in its functionalities, since it is developed in sync with `cardano-node`.

Reviewer: This list describes the advantages of the proposed cardano-api WASM. This list does not really belong in the context section.

Author: Addressed here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/8691367cfdfee2b5387249cda10827d54185edbe

> * **Robustness:** Because `cardano-api` is developed in Haskell, it relies heavily in type correctness, and because it is heavily tested, it is especially trustworth. Because the JavaScript/TypeScript API would take advantage of the same code, the robustness would be inherited.

Reviewer: ```suggestion
* **Robustness:** Because `cardano-api` is developed in Haskell, it relies heavily in type correctness, and because it is heavily tested, it is especially trustworthy. Because the JavaScript/TypeScript API would take advantage of the same code, the robustness would be inherited.
```

Author: Nice catch, suggestion applied here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/b83aa77b291e1090dd71bab258dba8e7dfac63eb

* **Pure browser support:** DApp libraries tend to be more focused on supporting execution in the backend (through Node.js and Bun). There are good reasons for doing this, but explicitly supporting JavaScript in the browser has some advantages:
  * **Incentivize adoption:** By making the library easier to try out, we would be reducing the friction to adoption.

>   * **Improved trust through decentralisation:** Since code is run in the browser, and thus the user's computer, it may be easier for the user to trust, even though transaction signing happens on the client side anyway.

Reviewer: I think this point isn't super clear to me. I understand that tx building and signing is done on the client side anyway?

Author: Addressed here: https://github.com/input-output-hk/cardano-node-wiki/commit/f212dd5693050f1881c396cf99a1058d73c3f37d

Other potential opportunities and considerations to keep in mind:

> * **Aditional conectivity:** We can potentially provide ways to connect to a `cardano-node` directly. And we can consider offering a provider that is not restricted by CORS, which would allow dApps run purely in the browser.

Reviewer: I think this needs a short explanation why CORS is a problem here, and how that provider solves the problem. Can be a footnote though.

Author: Yes, IOE or someone else, I don't know how or if it could be monetised. But I think it would improve the ecosystem in that it would allow dApps to run fully in the browser, thus improving transparency. I will rewrite to try to explain it better

* **Ease of Use & Documentation:** Most popular libraries provide introductory guides, that are comprehensive, simple, clear. Autocompletion in IDEs is crucial and often serves as de-facto documentation. In addition to all that, we should make comprehensive documentation for the whole API (reference) and make it easily accessible.

> * **CIP-30 Integration:** Existing libraries seem to provide wallet support through CIP-30. We may want to do that too, or at least make it easy for our API to interact with already existing libraries and APIs that already support CIP-30. The later may be hard because some of the most popular libraries integrate the wallets and transaction handling very closely and not necessarily in a functionaly pure way. It may also be used as a provider of information to some extent.

Reviewer: Link to CIP-30?

Author: Addressed here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/c265a9c2b910ed7a01b76d3614589ebd2c01423d

> * **Browser Compatibility & Testing:** For the reasons we mentioned before, we should use the CI to test compatibility with all potential workflows, including vanilla JavaScript and TypeScript from the browser, as well as integration with `Node.js`. Testing pure browser workflows can be tricky because we may need to use UI testing libraries like `Playwright`.

Reviewer: Cab you add a link to `Playwritght`?

Author: Addressed here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/c265a9c2b910ed7a01b76d3614589ebd2c01423d

* **Standard publication:** It is pretty standard for `Node.js` libraries to be distributed through channels like `npm`, so we should probably make sure to upload the library to those.

## Proposal

> We propose to create a new JavaScript and TypeScript library API for Cardano by compiling the `cardano-api` (Haskell) to WebAssembly (WASM) and adding a thing wrapper. This would allow developers to easy and cross-platform access to a subset of `cardano-api` functionalities from the browser and `Node.js` environments and similar.

Reviewer: ```suggestion
We propose to create a new JavaScript and TypeScript library API for Cardano by compiling the `cardano-api` (Haskell) to WebAssembly (WASM) and adding a thin wrapper. This would allow developers to easy and cross-platform access to a subset of `cardano-api` functionalities from the browser and `Node.js` environments and similar.
```

Author: Nice catch, suggestion applied here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/b83aa77b291e1090dd71bab258dba8e7dfac63eb

2.  **JavaScript/TypeScript interface:**
    * Provide simple, user-friendly, and well documented JavaScript ES6 module and TypeScript bindings.
    * Offer strong typing with TypeScript for enhanced robustness (types) and developer experience (autocompletion).
    * Support vanilla JavaScript usage for easy testing and adoption.
3.  **API design:**
    * **Low-Level access:** Expose a subset of `cardano-api` with a thin wrapper.
    * **High-Level abstractions (potential):** Consider optional, higher-level APIs, probably as separate modules. Examples:
        * A "virtual wallet" abstraction that handles UTXO management and coin selection automatically.
        * Simplified (possibly monadic) transaction building.
    * **CIP-30 support:** Potentially support CIP-30 compatibility to interact with browser wallet extensions, either as a separate module or as part of the "virtual wallet" abstraction.
    * **Data providers:** Allow configuration of blockchain data providers (e.g., BlockFrost, Koios). Consider offering a CORS-friendly proxy service (with robust caching to mitigate DDoS risk) to provide support for frontend-only DApp development.
4.  **Development Experience:**

>     * **Minimal JS glue:** Maximize Haskell code and minimize the JavaScript FFI glue code and try to generate glue code automatically if possible. This allows leveraging Haskell for most of the development, improving our productivity and the type safety of our code.

Reviewer: Code generation here sounds great!

    * **Error handling:** Ensure clear, descriptive, and catchable errors.
    * **BigInt for large amounts:** Use JavaScript `BigInt` for Ada/Lovelace amounts to prevent precision loss (also serialised/deserialised as/from strings).
5.  **Documentation & Examples:**
    * Comprehensive API documentation that is easy to navigate.
    * Step-by-step guides for getting started, including handling `async` calls.
    * Clear examples for main use cases.
    * Ensure good IDE autocompletion support (like in Visual Studio Code).
6.  **Distribution:** Distribute the library via NPM and potentially in `github.io`.
7.  **Testing:**
    * CI pipelines to consider for:
        * WASM compilation. (Build)
        * Generation and consistency of JS/TS bindings. (Like with golden tests)
        * Functionality across supported environments. (Browsers via tools like Playwright, and also Node.js)

>         * Glue generation verificaiton. (Like with golden tests)

Reviewer: ```suggestion
        * Glue generation verification. (Like with golden tests)
```

8.  **Stability:** Aim to maintain a stable API to avoid user frustration.

**Haskell to JavaScript Bridging Example:**

We would make a Haskell wrapper that is really close to what we want at the JavaScript level but using Haskell terms:

```haskell
signTransactionImpl :: Api.TxBody Api.ConwayEra -> Api.SigningKey Api.PaymentKey -> IO (Api.Tx Api.ConwayEra)
signTransactionImpl unsignedTx signingKey = do
    let sbe :: Api.ShelleyBasedEra Api.ConwayEra = Api.shelleyBasedEra
    let witness = Api.WitnessPaymentKey signingKey
    let oldApiSignedTx :: Api.Tx Api.ConwayEra = Api.signShelleyTransaction sbe unsignedTx [witness]
    return oldApiSignedTx
```

> Then we would write a conversion layer that is still in Haskell, but converts to and from JavaScript marshable types and usesw the pure Haskell function:

Reviewer: ```suggestion
Then we would write a conversion layer that is still in Haskell, but converts to and from JavaScript marshalable types and uses the pure Haskell function:
```

Author: Nice catch, suggestion applied here: https://github.com/input-output-hk/cardano-node-wiki/pull/75/commits/b83aa77b291e1090dd71bab258dba8e7dfac63eb

        * Simplified (possibly monadic) transaction building.
    * **CIP-30 support:** Potentially support CIP-30 compatibility to interact with browser wallet extensions, either as a separate module or as part of the "virtual wallet" abstraction.
    * **Data providers:** Allow configuration of blockchain data providers (e.g., [BlockFrost](https://blockfrost.io/), [Koios](https://koios.rest/)). Consider offering a CORS-friendly proxy service (with robust caching to mitigate DDoS risk) to provide support for frontend-only DApp development.
4.  **Development Experience:**
    * **Minimal JS glue:** Maximize Haskell code and minimize the JavaScript FFI glue code and try to generate glue code automatically if possible. This allows leveraging Haskell for most of the development, improving our productivity and the type safety of our code.
    * **Haskell API mirroring:** Because WASM ghc compiler target doesn't seem to support `haskell-language-server`, we should keep a pure Haskell API that closely mirrors the desired JavaScript API structure to be able to work on it using the HLS as far as possible, and use FFI just for type conversion.

>     * **Type conversion:** Implement `ToJSVal` / `FromJSVal` type-classes in Haskell, analogously to `ToJSON` / `FromJSON` in [Aeson](https://hackage.haskell.org/package/aeson), for conversions between Haskell and JS types. JSON string serialization can be used as an efficient intermediate representation (Haskell -> AESON -> String -> JS), and it can be used directly for simple types like addresses (and [Bech32](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) encoded data).

Author: I expanded the explanation here: https://github.com/input-output-hk/cardano-node-wiki/commit/7bcea874f43c569575aa8cb2997841d1818e2829

    -- Call the pure Haskell function
    oldApiSignedTx <- signTransactionImpl unsignedTx signingKey
    -- Convert Haskell type to JS
    let envelope = Api.serialiseToTextEnvelope (Just "Ledger Cddl Format") oldApiSignedTx
    jsonToJSVal envelope
```

This could be done more generically with a `FromJS` and `ToJS` class, as mentioned.

**JavaScript Initialization and Usage Example:**

The JavaScript glue can be very simple and basically packages the FFI Haskell functions:

```javascript
// cardano-api.js (or similar entry point for the library)

import { WASI } from "https://unpkg.com/@bjorn3/browser_wasi_shim@0.4.1/dist/index.js";
import ghc_wasm_jsffi from "./cardano-wasm.js"; // This is generated automatically from WASM
const __exports = {};
const wasi = new WASI([], [], []);

// Initialisation function returns a promise to an object with the API
async function initialize() {
  let {instance} = await WebAssembly.instantiateStreaming(fetch("./cardano-wasm.wasm"), {
    ghc_wasm_jsffi: ghc_wasm_jsffi(__exports),
    wasi_snapshot_preview1: wasi.wasiImport,
  })
  Object.assign(__exports, instance.exports);
  wasi.initialize(instance);

> 

Author: Yes, I think it is now done by the generated JSON. I think that because when I tried it the old way I couldn't get around calling `hs_init` manually, so I am relying on the fact that it works. Also, I am using [this](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/wasm.html) as a guide, which seems the most reliable explanation, and it doesn't talk about `hs_init` at all

export default initialize;
```

Example usage of the JS API could be as simple as defining an `async` function and then calling `.then()` on it:

```html
<html>
<body>
<script type="module">
  import cardano_api from "./cardano-api.js";
  let promise = cardano_api();
  async function do_async_work() {
    let api = await promise;

    let txIn = await api.mkTxIn("be6efd42a3d7b9a00d09d77a5d41e55ceaf0bd093a8aa8a893ce70d9caafd978", 0);
    console.log("Tx input:");
    console.log(txIn);

    let destAddr = "addr_test1vzpfxhjyjdlgk5c0xt8xw26avqxs52rtf69993j4tajehpcue4v2v";
    let privKey = "addr_sk1648253w4tf6fv5fk28dc7crsjsaw7d9ymhztd4favg3cwkhz7x8sl5u3ms";
    let amountInLovelace = 10_000_000n; // We pass Lovelace amounts as BigInt
    let feesInLovelace = 2_000_000n;

    let unsignedTx = await api.mkTransaction(txIn, destAddr, amountInLovelace, feesInLovelace);
    console.log("Tx body:");
    console.log(unsignedTx);

    let signedTx = await api.signTransaction(unsignedTx, privKey);
    console.log("Signed tx:");
    console.log(signedTx);
  }
  do_async_work().then(() => {});
</script>
</body>
</html>
```

**Stateful API (Wallet-like features):**

> If we do a stateful API (e.g., for a "virtual wallet"), the state can be managed in Haskell. Each function in the stateful API would take the state as its first parameter and return a tuple with the new state and the resulting state (this can be modeled as a [`State` monad](https://hackage.haskell.org/package/mtl/docs/Control-Monad-State-Class.html)). The JavaScript glue code would hold this state globally and pass it automatically for each function call, making it appear as an object-oriented API in JavaScript. This state could include loaded private keys (handled securely), UTXO sets, etc. This could be implemented as an abstraction layer that uses the core stateless functions under the hood.

Author: Addresed here: https://github.com/input-output-hk/cardano-node-wiki/commit/a0363ff968d084cae7e76168b300c2fc8e1ab431
